### ISSUE TITLE
Mattermost iOS "Connecting..."

### ISSUE THREAD

**Issue by asomervell** [mattermost/mattermost-mobile#742]
#### Summary
If I open the app, use it then navigate away, leaving it running in the background. Then re-open the app. When I view any group or direct message. I see a grey banner at the top saying "Connecting..."

I'm not sure if it is actually connecting still, and failing to load new messages? Or is it just saying connecting in the UI incorrectly?

#### Environment Information

- Device Name: iPhone 7 Plus
- OS Version: 11
- Mattermost App Version: 1.0 (Build 39)
- Mattermost Server Version: 3.10.0

#### Steps to reproduce
Open the app, navigate away to home screen, navigate back to app.

#### Expected behavior
"Connecting..." banner would go away, new messages would arrive.

#### Observed behavior
Screenshot attached.

#### Possible fixes
*Shrug

![Uploading IMG_4986.png…]()

**Comment by lfbrock** [comment]
Thanks @asomervell - I think the app is working okay, (if the connecting... goes away or turns green and says "connected" everything is working fine) but we should take another look at the user experience here. 

Basically when the app comes back from the background it needs to load the new messages (happens with all apps). 

We show the connecting bar while it's loading, but it seems like other apps show different types of loading screens or sometimes don't show one at all since it updates fairly quickly. So maybe we should explore other options here.

**Comment by enahum** [comment]
does it ever disappear? after the app is opened?

**Comment by enahum** [comment]
Well for people seeing the connecting bar all the time, please ensure that if your server uses nginx or any other webserver as a reverse proxy that it has support for upgrading the [websocket connection for the APIv4 endpoints](https://docs.mattermost.com/install/install-ubuntu-1604.html#configuring-nginx-as-a-proxy-for-mattermost-server)

**Comment by tobihagemann** [comment]
I'm experiencing the same issue. The "Connecting..." bar never disappears but everything else seem to work great.

We have installed a Mattermost server via Docker. Sorry, this might be a stupid question but: What exactly should we do? So yeah, I can see it's running nginx in the Dockerfile: https://github.com/mattermost/mattermost-docker/blob/master/web/Dockerfile

Shouldn't it "just work" when using Docker? :smile: Sooo basically this default file is "wrong"? https://github.com/mattermost/mattermost-docker/blob/master/web/mattermost

Is it possible to make it configurable?

**Comment by asomervell** [comment]
@enahum it doesn't go away, and I don't think it's loading new messages. Our instance is probably behind AWS ELB and a quick google says that doesn't "just work" so I might have to have a deeper look at that. If you've encountered that before or have docs please do LMK.

**Comment by cmayer-harris** [comment]
I'm seeing the same issue. I noticed that whenever I try to do anything on the IOS app, this log shows up:

```
[2017/07/18 20:36:51 UTC] [EROR] websocket connect err: websocket: origin not allowed
[2017/07/18 20:36:51 UTC] [EROR] /api/v4/websocket:connect code=500 rid=9u5q7b81epr69qyws9bxfjhi8r uid= ip= Failed to upgrade websocket connection [details: ]
```

If I set `"AllowCorsFrom": "*"`, the "Connecting..." bar goes away.

My instance is behind an AWS ALB, and using an nginx proxy.

**Comment by enahum** [comment]
@cmayer89 you should try and set the SiteURL in the config setting and that the URL used in the app matches that one.. and that should help you get rid of the `AllowCorsFrom: *`

**Comment by mrjfox** [comment]
Yes can confirm that setting "AllowCorsFrom": "*" in 'config/config.json' has resolved this issue as no longer seeing the error in the logs. Mattermost is running behind an AWS ALB.

**Comment by asomervell** [comment]
Confirm fixed, would be great to put that small bit in the docs :)

**Comment by cmayer89** [comment]
@enahum the SiteURL in my config.json file is already defined properly, as far as I can tell. It looks like this (but with my actual domain): 

```
{
    "ServiceSettings": {
        "SiteURL": "https://mattermost.example.com",
         ...
        "AllowCorsFrom": "*",
        ...
```

I agree it would be better not to have `"AllowCorsFrom": "*"` though. I'll continue to investigate and see if I can find a better solution.

Here is some more info about my setup:

Mattermost is running in a docker container (mattermost/mattermost-prod-app:4.0.0) on AWS ECS. I'm using another docker container as a reverse proxy (jwilder/nginx-proxy:latest). Traffic is routed to the proxy container via an AWS ALB. SSL is handled by the ALB (using an AWS ACM cert).

Here are the config files I'm using for my nginx proxy:
`/etc/nginx/proxy.conf`
```
# HTTP 1.1 support
proxy_http_version 1.1;
proxy_buffering off;
proxy_set_header Host $http_host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $proxy_connection;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
proxy_set_header X-Forwarded-Ssl $proxy_x_forwarded_ssl;
proxy_set_header X-Forwarded-Port $proxy_x_forwarded_port;

# Mitigate httpoxy attack (see README for details)
proxy_set_header Proxy "";

proxy_cache_path /var/cache/nginx levels=1:2 keys_zone=mattermost_cache:10m max_size=3g inactive=120m use_temp_path=off;
```

`/etc/nginx/vhost.d/mattermost.example.com`
```
location ~ /api/v[0-9]+/(users/)?websocket$ {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    client_max_body_size 50M;
    proxy_set_header Host $http_host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Frame-Options SAMEORIGIN;
    proxy_set_header X-Forwarded-Proto https;
    proxy_set_header X-Forwarded-Ssl on;
    proxy_buffers 256 16k;
    proxy_buffer_size 16k;
    proxy_read_timeout 600s;
    proxy_pass http://mattermost.example.com;
}
```

`/etc/nginx/vhost.d/mattermost.example.com_location`
```
client_max_body_size 50M;
proxy_set_header Connection "";
proxy_set_header Host $http_host;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto https;
proxy_set_header X-Forwarded-Ssl on;
proxy_set_header X-Frame-Options SAMEORIGIN;
proxy_buffers 256 16k;
proxy_buffer_size 16k;
proxy_read_timeout 600s;
proxy_cache mattermost_cache;
proxy_cache_revalidate on;
proxy_cache_min_uses 2;
proxy_cache_use_stale timeout;
proxy_cache_lock on;
```

**Comment by overheadhunter** [comment]
How exactly does the request to `wss://mattermost.example.com/api/v4/websocket` look like? What Host and Origin header does the iOS app send? I want to understand what difference there is between the app and a browser. Maybe it can be fixed in the app before opening the API to `*`.

**Comment by metasitenet** [comment]
Hi,

setting  "AllowCorsFrom": "*" parameter doesn't help while using Version: 3.10.0 (Team edition) and new native Mattermost **Android app** (version 1.0). "Connecting..." banner still visible.

**Comment by enahum** [comment]
@metasitenet anything in the server logs?

**Comment by metasitenet** [comment]
@enahum 

`[2017/07/24 08:28:17 EEST] [EROR] websocket connect err: websocket: not a websocket handshake: 'upgrade' token not found in 'Connection' header
[2017/07/24 08:28:17 EEST] [EROR] /api/v4/websocket:connect code=500 rid=7pinwwheibfpzdzuw1iia4muho uid=mh4rkin9u3fgmbs8ernu9xqade ip=83.188.113.160 Failed to upgrade websocket connection [details: ]
`

**Comment by enahum** [comment]
@metasitenet plase take a look at the docs https://docs.mattermost.com/install/troubleshooting.html#failed-to-upgrade-websocket-connection

**Comment by metasitenet** [comment]
Thanks @enahum , will look into it.

**Comment by enahum** [comment]
for anyone still having this issue you can take a look at issue #813

**Comment by gramakri** [comment]
We hit this with the Cloudron mattermost package (upstream bug: https://git.cloudron.io/cloudron/mattermost-app/issues/7).

From what I can tell, this doesn't appear to be a reverse proxy issue. The issue is that the Android and iOS app is not setting the ORIGIN header (I can see it is not set from the nginx logs). This means that the origin check in https://github.com/mattermost/platform/blob/master/utils/api.go#L16 fails.

**Comment by enahum** [comment]
Origin is being set here https://github.com/mattermost/mattermost-redux/blob/master/src/client/websocket_client.js#L86

**Comment by cmayer89** [comment]
This issue seems to be fixed for me now. I reverted the CORS settings in my config back to the original value, and everything seems to be working as expected.

```
"AllowCorsFrom": "",
```

App Version: 1.1.0 (Build 47)
Server Version: 4.1.0

**Comment by sadsfae** [comment]
Setting ```"AllowCorsFrom": "*"``` in ```config/config.json``` also fixed this for me.  I'm running Apache frontend doing reverse proxy duties.

**Comment by overheadhunter** [comment]
Am I the only one concerned about security here?

If you guys all set your CORS policy to "allow requests from anywhere", you're effectively inviting people to write phishing websites to attack your Mattermost servers.

This needs to be fixed in the mobile apps by sending the appropriate Origin header.

**Comment by lfbrock** [comment]
Hi @overheadhunter - you typically shouldn't need to set `"AllowCorsFrom": "*"`. 

In the mobile apps, we do set the origin header in the websocket connection using the value the person entered as the server URL. If people are having issues, it seems like it may be related to their specific SiteURL setting or another config value - if anyone can provide more information on their set up, it would be helpful.

**Comment by jsg2021** [comment]
We moved our instance to a public ip instead of behind other things, the app works much better now.

**Comment by riemers** [comment]
I have a related issue. I can login, i can see the channels and all the text. I just cannot send any messages. I get a ! mark to try again, if i just use a browser (Safari) from my phone it works just fine. Also get the "no internet" message at the top, i can press the spinner but nothing seems to happen.

**Comment by lfbrock** [comment]
Hi @riemers - sorry just saw this message, if you're still seeing the issue can you try: 

1. Upgrade to the latest version of the app (v1.4)
2. Go to settings menu > Report a problem, and copy any errors you see from there into a new GitHub issue?

**Comment by lfbrock** [comment]
Closing this issue - if anyone is still having problems, can you open a new GitHub issue with: 

1. The errors you're seeing
2. Your server config settings, and NGINX config settings

**Comment by riemers** [comment]
I did do the upgrades and it seems its doing much better now, only have to reconnect if idle long (but that can be due to mobile iron which sits in between)

**Comment by alecisec** [comment]
I'm also having this issue with the 1.4.1 client on IOS. 
Android 1.4.1 works perfectly, IOS Classic 3.10.1 works fine.

Our mattermost server sits behind an AWS ALB

**Comment by alecisec** [comment]
I resolved this by removing the port number from the "Site URL" 
it was previously - 

https://site.com:443
and I changed it to 
https://site.com

this stopped the IOS react native application from failing due to incorrect websocket origins. 
Setting allow CORS to * ( as @lfbrock stated) is not a good idea.

while the config states:

The URL that users will use to access Mattermost. Standard ports, such as 80 and 443, can be omitted, but non-standard ports are required. For example: http://mattermost.example.com:8065. This setting is required.

it seems that stating a standard port here breaks this one particular app's connectivity
